### PR TITLE
support upper layer customize pipeline namespace

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
support upper layer customize pipeline namespace

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/all?id=58451&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls1MDZdLCJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASKe)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  support customize namespace            |
| 🇨🇳 中文    |   支持上层自定义流水线命名空间           |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: apistructs/job.go [hunk 1]
```diff
@@ -66,30 +66,31 @@ type Job struct {
...
-	ClusterName    string                 `json:"clusterName,omitempty"`
-	Kind           string                 `json:"kind"`              // Metronome/FLink/Spark/LocalDocker/Swarm/Kubernetes
-	Depends        []string               `json:"depends,omitempty"` // JobName
-	PreFetcher     *PreFetcher            `json:"preFetcher,omitempty"`
-	BackoffLimit   int                    `json:"backoffLimit,omitempty"`
-	Params         map[string]interface{} `json:"params,omitempty"`
-	TaskContainers []TaskContainer        `json:"taskContainers"`
+	Name              string                 `json:"name"`
+	Namespace         string                 `json:"namespace"`         // the default namespace is "default"
+	IsCustomNamespace bool                   `json:"isCustomNamespace"` // if isCustomNamespace is true, resource gc should be attention don't delete namespace
```
Comment (sfwn) [inline apistructs/job.go:71]: More clearly.
`IsPipelineControlledNs`.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
How about other executor's namespace gc logic?

**Comment by chengjoey** [comment]
> How about other executor's namespace gc logic?

add delete ns Judgment logic in k8sflink `create` func and k8sspark

File: apistructs/job.go [hunk 2]
```diff
@@ -66,30 +66,31 @@ type Job struct {
...
-	ClusterName    string                 `json:"clusterName,omitempty"`
-	Kind           string                 `json:"kind"`              // Metronome/FLink/Spark/LocalDocker/Swarm/Kubernetes
-	Depends        []string               `json:"depends,omitempty"` // JobName
-	PreFetcher     *PreFetcher            `json:"preFetcher,omitempty"`
-	BackoffLimit   int                    `json:"backoffLimit,omitempty"`
-	Params         map[string]interface{} `json:"params,omitempty"`
-	TaskContainers []TaskContainer        `json:"taskContainers"`
+	Name                   string                 `json:"name"`
+	Namespace              string                 `json:"namespace"`              // the default namespace is "default"
+	IsPipelineControlledNs bool                   `json:"IsPipelineControlledNs"` // if IsPipelineControlledNs is true, resource gc should be attention don't delete namespace
```
Comment (sfwn) [inline apistructs/job.go:71]: if IsPipelineControlledNs is true => false

Comment (sfwn) [reply apistructs/job.go:71]: Old existed records' `IsPipelineControlledNs` is default false, will cause gc skipped.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/job.go (hunk 2 above)
Comment (sfwn) [reply apistructs/job.go:71]: Change to `NotPipelineControlledNs`

**Review by sfwn — APPROVED** [review]

**Pull request merged**